### PR TITLE
[suspendplugin] Reduce power consumption for wmtWifi driver. Contributes JB#49168

### PR DESCRIPTION
**Pull request by ikamaletdinov**
Some devices with gen3 driver has specific behaviour: to suspend it correctly power_save mode should be turned off and then turned on after connecting to WiFi network.

### CODE REVIEW

File: src/suspendplugin.c [hunk 1]
```diff
@@ -384,6 +441,14 @@ suspend_handle_display_off()
...
         // only suspend/unsuspend if we are not in ap mode
         if ((if_nametoindex("ap0")) == 0) {
+            /*
+                wmtWifi (with gen3 driver at least) in Android requires to set power_save
+                to `off` and after that power_save to `on` to correctly suspend wmtWifi device.
+                We are doing that just before calling suspend_set_wmtwifi() after display is turned off.
+                This solution doesn't give us any disadvantages in comparison with Android behaviour.
+            */
+            (void) suspend_set_powersave("wlan0", FALSE);
+            (void) suspend_set_powersave("wlan0", TRUE);
```
Comment (krnlyng) [inline src/suspendplugin.c]: Does it make sense to declare the function with a return value but ignore the return value every time it is being used?

File: src/suspendplugin.c [hunk 2]
```diff
@@ -238,6 +238,63 @@ suspend_set_wowlan(
...
+    int err = 0;
+    struct nl_msg *msg;
+
+    enum nl80211_ps_state ps_state;
+
+    int ifindex = 0;
+
+    ifindex = if_nametoindex(ifname);
+
+    if (ifindex == 0) {
```
Comment (krnlyng) [inline src/suspendplugin.c:261]: This same sequence is used in many places, should we turn it into a macro or function?

File: src/suspendplugin.c (hunk 1 above)
Comment (ikamaletdinov) [reply src/suspendplugin.c]: I'm sure that return value could be used later by us or other people. Just `reserved` return value.

File: src/suspendplugin.c (hunk 2 above)
Comment (ikamaletdinov) [reply src/suspendplugin.c:261]: I'm really not sure about that. To be honest I think that we should refactor that plugin as soon as we will have more devices which we are supporting using that plugin. Lets keep it as is for now.

File: src/suspendplugin.c (hunk 1 above)
Comment (monich) [reply src/suspendplugin.c]: I'm not so sure about that :) Besides, it's easy enough to add when and if it becomes needed. I suggest to stick to "do not introduce new entities without necessity" principle. And this `(void)` thing isn't necessary too, it's just littering  the code IMO. Not that it's a big deal but if it were the code that I'm maintaining, I would take it out.

File: src/suspendplugin.c (hunk 2 above)
Comment (ikamaletdinov) [reply src/suspendplugin.c:261]: @monich @krnlyng  Can we keep it as is? 

I'm really not sure that part of code listed above should be moved to function or/macro.

File: src/suspendplugin.c [hunk 3]
```diff
@@ -238,6 +238,61 @@ suspend_set_wowlan(
...
+
+    DBG("iface %s, setting powersave.", ifname);
+
+    msg = nlmsg_alloc();
+
+    genlmsg_put(msg, 0, 0, driver_id, 0, 0, NL80211_CMD_SET_POWER_SAVE, 0);
+
+    nla_put_u32(msg, NL80211_ATTR_IFINDEX, ifindex);
+
+    if(is_enable)
```
Comment (krnlyng) [inline src/suspendplugin.c]: nitpick: there is a space missing between "if" and "("

**Review by krnlyng — APPROVED** [review]
Unless somebody else (@monich ?) still has some objections to this i think it is fine.

Though i would prefer if we could do some cleanup to remove some of the code duplication.

File: src/suspendplugin.c (hunk 3 above)
Comment (krnlyng) [reply src/suspendplugin.c]: can you fix this? then i'll merge.

Comment (ikamaletdinov) [reply src/suspendplugin.c]: sure, sorry I've fixed same issue in another repo and forget about this repo

**Review by krnlyng — APPROVED** [review]

**Pull request merged**